gmi_egads/gmi_egads.c

```diff
@@ -0,0 +1,832 @@
+static void bbox(struct gmi_model* m,
+                 struct gmi_ent* e,
+                 double bmin[3],
+                 double bmax[3])
+{
+  (void)m;
+  double box[6];
+  egads_ent *eg_ent = (egads_ent*)e;
+  ego ego_ent = eg_ent->ego_ent;
+
+  if (eg_ent->ego_ent == NULL)
+  {
+    get_3D_bounding_box(eg_ent, box);
+  }
+  else
+  {
+    EG_getBoundingBox(ego_ent, box);
+  }
+  bmin[0] = box[0];
+  bmin[1] = box[1];
+  bmin[2] = box[2];
+  bmax[0] = box[3];
+  bmax[1] = box[4];
+  bmax[2] = box[5];
+}
+
+/// For any given vertex, edge, or face e, this function can be used
+/// to see if the e is in the closure of entity et.
+static int is_in_closure_of(struct gmi_model* m,
+                            struct gmi_ent* e,
+                            struct gmi_ent* et)
+{
+  egads_model *egm = (egads_model*)m;
+
+  egads_ent *eg_ent = (egads_ent*)e;
+  ego ego_ent = eg_ent->ego_ent;
+
+  egads_ent *eg_region = (egads_ent*)et;
+  ego ego_region = eg_ent->ego_ent;
+
+  int num_adjacent = 0;
+
+  int ent_dim = get_dim(m, e);
+  int region_dim = get_dim(m, et);
+  if (ent_dim == 3 || region_dim == 3)
+  {
+    egads_ent **adjacent_ents;
+    // get entities of dim ent_dim adjacent to eg_region (will be downward adjacent)
+    get_3D_adjacency(m, eg_region, ent_dim, &num_adjacent, &adjacent_ents);
+
+    for (int i = 0; i < num_adjacent; i++)
+    {
+      if (adjacent_ents[i]->tag == eg_ent->tag)
+      {
+        EG_free(adjacent_ents);
+        return 1;
+      }
+    }
+    EG_free(adjacent_ents);
+    return 0;
+  }
+  else
+  {
+    ego *adjacent_egos = NULL;
+    if (ent_dim == 0)
+      EG_getBodyTopos(egm->eg_body, ego_region, NODE, &num_adjacent, &adjacent_egos);
+    else if (ent_dim == 1)
+      EG_getBodyTopos(egm->eg_body, ego_region, EDGE, &num_adjacent, &adjacent_egos);
+    else if (ent_dim == 2)
+      EG_getBodyTopos(egm->eg_body, ego_region, FACE, &num_adjacent, &adjacent_egos);
+    for (int i = 0; i < num_adjacent; ++i)
+    {
+      if (EG_isEquivalent(ego_ent, adjacent_egos[i]))
+      {
+        EG_free(adjacent_egos);
+        return 1;
+      }
+    }
+    EG_free(adjacent_egos);
+    return 0;
+  }
+}
+
+/// what does this function do?
+static int is_discrete_ent(struct gmi_model* m, struct gmi_ent* e)
+{
+  (void)m;
+  (void)e;
+  gmi_fail("is_discrete_ent not implemented");
+}
+
+static void destroy(struct gmi_model* m)
+{
+  egads_model *egm = (egads_model*)m;
+
+  for (int dim = 0; dim < 4; ++dim)
+  {
+    EG_free(egm->egads_model_ents[dim]);
+  }
+
+  int sizes[] = {m->n[3], m->n[3], m->n[3],
+                 m->n[0], m->n[1], m->n[2]};
+  for (int i = 0; i < 6; ++i)
+  {
+    for (int j = 0; j < sizes[i]; ++j)
+    {
+      EG_free(egm->adjacency_table[i][j]);
+    }
+    EG_free(egm->adjacency_table[i]);
+  }
+
+  free(egm);
+}
+
+static struct gmi_model_ops ops;
+
+/// TODO: Come up with a better flag? 
+/// TODO: re-write for EGADSlite - model loading is different
+// #ifdef HAVE_EGADS
+#if 1
```

Comment:
IIUC, the generated cmake in this PR provides a preprocessor macro indicating if egads lite is available.

Reply:
This is outdated. I rewrote this for EGADSlite in the most recent commit.